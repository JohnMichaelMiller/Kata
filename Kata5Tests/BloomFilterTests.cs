using Microsoft.VisualStudio.TestTools.UnitTesting;
using DifferentialAbstraction;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using ApprovalTests;
using ApprovalTests.Reporters;
using ApprovalUtilities.Utilities;

namespace DifferentialAbstraction.Tests
{
    [TestClass()]
    public class BloomFilterTests
    {
        const string fileName = @"C:\Users\John\Source\Repos\Kata\Kata5\wordlist.txt";
        private readonly BloomFilter _bfilter;

        public BloomFilter Bfilter
        {
            get { return _bfilter; }
        }

        public BloomFilterTests()
        {
            Dictionary<string, int> words = new WordHash(fileName);
            _bfilter = new BloomFilter(words);
        }

        [TestMethod()]
        public void EqualBitArraysShouldCompareAsEqual()
        {
            BitArray wordBA = new BitArray(10);
            BitArray bloomBA = new BitArray(10);
            bool result = false;

            wordBA[0] = true;
            wordBA[1] = true;
            wordBA[2] = false;
            wordBA[3] = true;
            wordBA[4] = true;
            wordBA[5] = false;
            wordBA[6] = false;
            wordBA[7] = true;
            wordBA[8] = false;
            wordBA[9] = true;

            bloomBA = wordBA;

            result = Bfilter.CompareWordToBloom(wordBA, bloomBA);
            Assert.IsTrue(result);
        }

        [TestMethod()]
        public void WhenWordBitsAreFalseAndBloomBitsAreTrueTheyShouldCompareAsEqual()
        {
            BitArray wordBA = new BitArray(10);
            bool result = false;

            wordBA[0] = true;
            wordBA[1] = true;
            wordBA[2] = false;
            wordBA[3] = true;
            wordBA[4] = true;
            wordBA[5] = false;
            wordBA[6] = false;
            wordBA[7] = true;
            wordBA[8] = false;
            wordBA[9] = true;

            var bloomBA = wordBA;

            bloomBA[5] = true;

            result = Bfilter.CompareWordToBloom(wordBA, bloomBA);
            Assert.IsTrue(result);
        }

        [TestMethod()]
        public void WhenWordBitsAreTrueAndBloomBitsAreFalseTheyShouldCompareAsNotEqual()
        {
            BitArray wordBA = new BitArray(10);
            bool result = false;

            wordBA[0] = true;
            wordBA[1] = true;
            wordBA[2] = false;
            wordBA[3] = true;
            wordBA[4] = true;
            wordBA[5] = false;
            wordBA[6] = false;
            wordBA[7] = true;
            wordBA[8] = false;
            wordBA[9] = true;

            var bloomBA = wordBA;

            bloomBA[1] = false;

            result = Bfilter.CompareWordToBloom(wordBA, bloomBA);
            Assert.IsTrue(result);
        }

        [TestMethod()]
        public void MergeArraysTest()
        {
            BitArray word = new BitArray(new int[] {128});
            Bfilter.MergeWordWiithBits(word);
        }

        [TestMethod()]
        public void FindWords()
        {
            Assert.IsTrue(Bfilter.FindWord("A"), "bm.FindWord('A') should be true and was not.");
            Assert.IsTrue(Bfilter.FindWord("AA"), "bm.FindWord('AA') should be true and was not.");
            Assert.IsFalse(Bfilter.FindWord("B"), "bm.FindWord('B') should be false and was not. ");
            Assert.IsFalse(Bfilter.FindWord("C"), "bm.FindWord('C') should be false and was not. ");
            Assert.IsFalse(Bfilter.FindWord("D"), "bm.FindWord('D') should be false and was not. ");
            Assert.IsFalse(Bfilter.FindWord("E"), "bm.FindWord('E') should be false and was not. ");

        }

        [TestMethod()]
        [UseReporter(typeof(DiffReporter))]
        public void BloomFilterTest()
        {
            var result = new bool[Bfilter.Bits.Count];
            Bfilter.Bits.CopyTo(result, 0);
            Approvals.VerifyAll(result, "Bits");
        }

    }
}