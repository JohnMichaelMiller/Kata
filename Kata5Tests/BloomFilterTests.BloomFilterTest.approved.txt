Bits[0] = True
Bits[1] = True
Bits[2] = True
Bits[3] = True
Bits[4] = True
Bits[5] = True
Bits[6] = True
Bits[7] = True
Bits[8] = True
Bits[9] = True
Bits[10] = True
Bits[11] = True
Bits[12] = True
Bits[13] = True
Bits[14] = True
Bits[15] = True
Bits[16] = True
Bits[17] = True
Bits[18] = True
Bits[19] = True
Bits[20] = True
Bits[21] = True
Bits[22] = True
Bits[23] = True
Bits[24] = True
Bits[25] = True
Bits[26] = True
Bits[27] = True
Bits[28] = True
Bits[29] = True
Bits[30] = True
Bits[31] = True
